# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd

df = pd.read_csv('./data/supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [2]:
df.sort_values(by='Unit price', ascending=False).head()

Invoice ID Branch       City Customer type Gender         Product line  \
983  148-41-7930      C  Naypyitaw        Normal   Male    Health and beauty   
122  219-22-9386      B   Mandalay        Member   Male    Sports and travel   
930  641-62-7288      B   Mandalay        Normal   Male   Home and lifestyle   
494  437-53-3084      B   Mandalay        Normal   Male  Fashion accessories   
283  667-92-0055      A     Yangon        Member   Male    Health and beauty   

     Unit price  Quantity  Tax 5%    Total       Date   Time      Payment  \
983       99.96         7  34.986  734.706  1/23/2019  10:33         Cash   
122       99.96         9  44.982  944.622   3/9/2019  17:26  Credit card   
930       99.92         6  29.976  629.496  3/24/2019  13:33      Ewallet   
494       99.89         2   9.989  209.769  2/26/2019  11:48      Ewallet   
283       99.83         6  29.949  628.929   3/4/2019  15:02      Ewallet   

       cogs  gross margin percentage  gross income  Rating  
983  699.72                 4.761905        34.986     6.1  
122  899.64                 4.761905        44.982     4.2  
930  599.52                 4.761905        29.976     7.1  
494  199.78                 4.761905         9.989     7.1  
283  598.98                 4.761905        29.949     8.5

2. Знайдіть інвойс (рядок в даних) в категорії товарів Electronic accessories, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [3]:
df[df['Product line'] == 'Electronic accessories'].sort_values(by=['Quantity', 'Unit price'], ascending=False).head(1)

Invoice ID Branch    City Customer type Gender            Product line  \
105  704-48-3927      A  Yangon        Member   Male  Electronic accessories   

     Unit price  Quantity  Tax 5%    Total       Date   Time  Payment   cogs  \
105       88.67        10  44.335  931.035  1/12/2019  14:50  Ewallet  886.7   

     gross margin percentage  gross income  Rating  
105                 4.761905        44.335     7.3

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу: високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 до 8 не включаючи - то він середній (`Medium`)
- інакше - низький (`Low`).

In [5]:
def rating_category(rating_value):
    if rating_value >= 8:
        return 'High'
    elif 6 <= rating_value < 8:
        return 'Medium'
    else:
        return 'Low'

df['Rating Category'] = df.Rating.apply(rating_category)
df['Rating Category'].head(3)


0      High
1      High
2    Medium
Name: Rating Category, dtype: object

3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [6]:
df['Rating Category'].value_counts(normalize=True)

Medium    0.356
High      0.329
Low       0.315
Name: Rating Category, dtype: float64

4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [11]:
time = '11:28'

In [14]:
def get_hour(time):
    return int(time.split(':')[0])

get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [16]:
df['Hour'] = df.Time.apply(get_hour)
df['Hour'].head(3)

0    13
1    10
2    13
Name: Hour, dtype: int64

4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому, на Ваш погляд, саме о цих годинах? Чи це логічно, що найбільше покупок саме в ці години?".

In [19]:
df['Hour'].value_counts().sort_values(ascending=False)

19    113
13    103
15    102
10    101
18     93
11     90
12     89
14     83
16     77
20     75
17     74
Name: Hour, dtype: int64

Найбільше покупок здійснюють о 19 (113 покупок) та 13 (103 покупки) годині. Здається логічним, що найбльша кількість покупок припадає на завершення робочого дня та обідню перерву.
О 15 години (102 покупки), можливо, закінчується 1 зміна. 
О 10 години (101 покупка) - приходять на відкриття або до роботи. 

5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [24]:
df.groupby('City')[['Total']].mean().rename(columns={'Total': 'Total mean'})

Total mean
City                 
Mandalay   319.872506
Naypyitaw  337.099715
Yangon     312.354031

6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [26]:
df.groupby('City')[['Total']].agg(['count', 'mean', 'min', 'max'])

Total                              
          count        mean      min      max
City                                         
Mandalay    332  319.872506  18.6375  1022.49
Naypyitaw   328  337.099715  10.6785  1042.65
Yangon      340  312.354031  12.6945  1039.29

Найбільше середнє значення для загальної суми чеку в місті `Naypyitaw`, причому у цьому місті найменша кулькість покупок, однак найбільший розмах загальної суми чеку (найменше значення чеку та найбільше значення чеку також у цьому місті). 

Можна рекомендувати мережі супермаркетів розширити асортимент товарів, збільшити кількість дешевих товарів. Однак дане припущення потребує додаткової перевірки.

7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [27]:
df.to_csv('./data/supermarket_sales_enriched.csv', index=False)